Describe L2 blocks and L1 batches (#96)
matter-labs/zksync-web-v2-docs: docs/dev/zksync-v2/blocks-and-time.md

# Block numbers and time

On ZkSync there are two types of blocks: L2 blocks and L1 batches.

## L2 blocks

L2 block is similar to a common block on L1.
L2 block time is almost constant, like a few seconds.

> In this documentation by default the word "block" actually

Reviewer: Lines seem to be unnecessarily short.

means "L2 block" including API.
L2 block hash is deterministic
and is basically equal to `keccak256(bytes4(l2_block_number))`.

## L1 batches

L1 batch is a range of L2 blocks that are committed
and finalized on L1 as one whole.

> L1 batch time can be different: from minutes to several hours

Reviewer: "depending on the system load" is an obscure statement. I think you should clarify that to achieve cost optimization we try use the whole capacity we have, and generally we only try to seal the batch when it's full, whereas "cheap" l2 blocks can be sealed quickly.

# Block numbers and time

> On ZkSync there are two types of blocks: L2 blocks and L1 batches.

Reviewer: Why do we need 2 types? Why the latter called "batches" and not "blocks" (though they're called blocks at the beginning of the sentence)?

## L2 blocks

> L2 block is similar to a common block on L1. L2 block time is almost constant, like a few seconds.

Reviewer: What is a "common block"? Does "uncommon block" exist on L1?
"Almost constant, like a few seconds" sounds very uncertain -- is it a formal definition? Why can't it be "fully constant"?

> In this documentation by default the word "block" actually means "L2 block" including API.

Reviewer: "word ... actually means" makes it sound like we wrote the documents incorrectly and are unable to use the correct term outselves.
To make it sound better we need to work on phrasing, i.e. "For simplicity, when we mention "block" in this documentation, "L2 block" is always assumed, unless specified otherwise".
"including API" part is a bit confusing to me -- what's so special about API?

> In this documentation by default the word "block" actually means "L2 block" including API.

Reviewer 2: I think this may is simpler:

```suggestion
In this documentation and zkSync API, the word "block" refers to the "L2 block".
```

> L2 block hash is deterministic and is basically equal to `keccak256(bytes4(l2_block_number))`.

Reviewer 2: Maybe let's have it as `keccak256(l2_block_number)`?

IMHO does not really matter, but the `bytes4` part complicates the thing a bit. Also, formally, there is a difference between 

```
keccak256(bytes.concat(bytes4(l2_block_number)))
```

and 

```
keccak256(bytes.concat(bytes32(uint256(l2_block_number))))
```

The former is left-aligned, while the latter is right-aligned. We don't need to elaborate that much in the documentation, but I am still curious :)

## L1 batches

> L1 batch is a range of L2 blocks that are committed and finalized on L1 as one whole.

Reviewer 2: > committed and finalized

Just in case the users don't know what committing and finalising means, please add a link to the page, where we explain how ZK rollups work in general

> The purpose of introducing L1 batch conception is to achieve cost optimization by using the whole capacity we have.

Reviewer: What does "whole capacity" mean? What defines said capacity?

> Thus, L1 batch is sealed only if its capacity limit, or a timeout is reached.

Reviewer: Why do we want to have a timeout if we want to utilize the whole capacity?

That is why L1 batch time can be different: from minutes under the high load to several hours when the batch is closed by timeout.

Note, that it does not relate to L2 blocks, they are "cheap" and can be sealed fast.
For users, monitoring L1 batches might be interesting only in case they are waiting for finality on L1,

> for all other cases they should use L2 blocks.

Reviewer: How can they "use" L2 blocks? Is it opt-in?

> ## `block` properties

Reviewer 2: There should always be a newline between a title and the content

```suggestion
## `block` properties

```

> While executing code block properties (`block.number`, `block.hash`, `block.timestamp`, etc) are actually the properties of the current L1 batch.

Reviewer 2: ```suggestion
While executing code block properties (`block.number`, `block.hash`, `block.timestamp`, etc) the properties of the current L1 batch are returned.
```

> So, you shouldn't rely on the fact that they are changing every few seconds.

Reviewer: This is confusing and seemingly incorrect. If these are properties of L1 batch, you *can* rely on them, no?
You should not rely on the hash of L2 block, since it's deterministic. And there should be clarification of what does "rely" mean in this context.

> So, you shouldn't rely on the fact that they are changing every few seconds.

Reviewer 2: ```suggestion
That is why you shouldn't rely on the fact that they are changing every 10-15 seconds.
```
